qubes/mgmt.py

```diff
@@ -534,3 +570,95 @@ def vm_kill(self):
         assert not self.arg
         self.fire_event_for_permission()
         yield from self.dest.kill()
+
+    @api('mgmt.Events', no_payload=True)
+    @asyncio.coroutine
+    def events(self):
+        assert not self.arg
+
+        # run until client connection is terminated
+        self.cancellable = True
+        wait_for_cancel = asyncio.get_event_loop().create_future()
+
+        # cache event filters, to not call an event each time an event arrives
+        event_filters = self.fire_event_for_permission()
+
+        def handler(subject, event, **kwargs):
+            if self.dest.name != 'dom0' and subject != self.dest:
+                return
+            if event.startswith('mgmt-permission:'):
+                return
+            for selector in event_filters:
+                if not selector((subject, event, kwargs)):
+                    return
+            self.send_event(subject, event, **kwargs)
+
+        if self.dest.name == 'dom0':
+            type(self.app).add_handler('*', handler)
+        qubes.vm.BaseVM.add_handler('*', handler)
+
+        # send artificial event as a confirmation that connection is established
+        self.send_event(self.app, 'connection-established')
+
+        try:
+            yield from wait_for_cancel
+        except asyncio.CancelledError:
```

Comment:
This seems suspicious, because technically ``CancelledError`` may have resulted from either ``self.cancel()`` or ``wait_for_cancel.cancel()``.  Those conditions aren't distinguishable from try/except.

I don't currently know, how this one should be expressed. I'd have to read some other people's code.